chain-signatures/node/src/protocol/triple.rs

```diff
@@ -93,12 +254,181 @@ pub struct TripleManager {
+                let participants = participants.keys_vec();
+                let generator = e.insert(TripleGenerator::new(
+                    me,
+                    threshold,
+                    id,
+                    participants,
+                    cfg.triple.generation_timeout,
+                )?);
+                self.queued.push_back(id);
+                crate::metrics::NUM_TOTAL_HISTORICAL_TRIPLE_GENERATORS
+                    .with_label_values(&[my_account_id.as_str()])
+                    .inc();
+                Ok(Some(generator.clone()))
+            }
+            Entry::Occupied(e) => Ok(Some(e.get().clone())),
+        }
+    }
+
+    pub async fn poke(
+        &mut self,
+        me: Participant,
+        my_account_id: &AccountId,
+        epoch: u64,
+        cfg: &ProtocolConfig,
+    ) -> (
+        Vec<(Triple, bool)>,
+        Vec<(Participant, TripleMessage)>,
+        HashMap<TripleId, ProtocolError>,
+    ) {
+        // Add more protocols to the ongoing pool if there is space.
+        let to_generate_len = cfg.max_concurrent_generation as usize - self.ongoing.len();
+        if !self.queued.is_empty() && to_generate_len > 0 {
+            for _ in 0..to_generate_len {
+                if let Some(id) = self.queued.pop_front() {
+                    self.ongoing.insert(id);
+                    let generator = self.generators.get(&id).unwrap();
+                    self.ongoing_tasks
+                        .push_back((id, generator.spawn_execution(me, my_account_id, epoch)));
+                }
+            }
+        }
+
+        // spawn these tasks again if they already completed with Action::Wait:
+        for id in &self.ongoing {
+            if !self
+                .ongoing_tasks
+                .iter()
+                .any(|(running_id, _)| running_id == id)
+            {
+                let generator = self.generators.get(&id).unwrap();
+                self.ongoing_tasks
+                    .push_back((*id, generator.spawn_execution(me, my_account_id, epoch)));
+            }
+        }
+
+        let mut triples = Vec::new();
+        let mut messages = Vec::new();
+        let mut errors = HashMap::new();
+
+        let mut interval = tokio::time::interval(Duration::from_millis(5));
+        let started = Instant::now();
+
+        // Go through each running task and see if it's done. This will apply a protocol_budget which will
+        // yield back control to the main loop if the time is up. If it is done, remove it from the ongoing_tasks.
+        // If the TripleGenerator is not done after this, a new task will be spawned in the next iteration
+        // in the case that the TripleGenerator is waiting.
+        while let Some((id, task)) = self.ongoing_tasks.pop_front() {
+            interval.tick().await;
+            if started.elapsed() > self.protocol_budget {
+                self.ongoing_tasks.push_back((id, task));
+                break;
```

Comment:
because we used up all our protocol budget. The whole of triple tasks is only given 100ms by default. If it goes beyond that, yield back control to the main loop. This does not stop all the running tasks from completing but just that we're no longer waiting for them to complete